docs/sdg/sdg-api-interface.md

```diff
@@ -0,0 +1,71 @@
+# SDG Library Design
+
+
+## Objective
+Library called `instructlab-sdg` that can be called per seed example that includes question and answer pairs, and context for grounded skills.
+
+## Structure of the SDG Library
+
+We propose the following structure for the SDG library. There will be config files that contain all the prompt templates for the pipelines. 
+
+
+```
+- src/instructlab/sdg/
+    - configs/ 
+        - gen_q.yaml 
+        - gen_a.yaml
+        - ...
+    - init.py 
+    - block.py 
+    - llmblock.py 
+    - pipeline.py 
+    - sdg.py
+```
+
+
+## CLI
+
+The CLI client uses the instructlab SDG library and provides it a run configuration with input parameters.
+
+
+```
+# cli_driver.py
+
+import sdg
+
+client = openai_client(endpoint)
+cli_sdg = sdg(run_config, client)  # run config has all the variables like num_samples, pipelinesteps etc
+generated_samples = cli_sdg.generate()
+```
+
+The run configuration contains the parameters needed to run the SDG code library as well as points to the templates needed to run the SDG code as well as the prompt template and the default model system prompt template.
+
+* num_samples is the number of synthetic samples that you wish to generate per seed example
+* max_retry is the maximum number of non-greedy retries you want to make if the num_samples is not reached. The number of samples in the generated output will be the samples achieved until max_retry is reached
+* Pipeline steps contains the steps that you want to invoke in the SDG pipeline and the prompt configurations per step. The variable names of the blocks can be anything and the prompt configs must be compatible with the teacher model
+*  max_new_tokens is the maximum number of tokens we want to generate. In other words, the size of the output sequence, not including the tokens in the prompt. 
+*  model_name is the teacher model we would want to use to generate the synthetic data.
+*  model_template and stop_token are the paramaters for generation template.
+
+```
+# run_config.yaml
```

Comment:
Does this need to be a yaml file?

or is it actually a Python dict passed to the library?

I actually think it would be a nice simplification to not worry about config files at all, and from the library perspective, assume configuration is passed in via data structures.

How that config is constructed could be a problem of the library consumer.  Maybe they hardcode it.  maybe they allow a subset to be considered.  Some could be driven by CLI args, for example.